### ISSUE TITLE
SSE support

### ISSUE THREAD

**Issue by andrewfg** [mockoon/mockoon#990]
### Issue
Old style REST API devices have GET/PUT commands whereby the client can read/write data to the server (i.e. a client driven, polling mechanism). But nowadays many REST API devices also support Server Sent Eventing (SSE) whereby the server can inform the client directly if anything changes. (i.e. a server event driven mechanism).

### Example
This is often the case in home automation devices (e.g. [Philips Hue](https://developers.meethue.com/develop/hue-api-v2/)) where the server device can (say) notify the client that a light was turned on or a temperature has changed.

### Analog
Some implementers use SSE to send event notifications. But some use WebSockets instead. The implementation is slightly different, but the purpose is the same. => So see #83 

### Proposed Solution
Mockoon should add the following features..
- Ability for a client to open an SSE connection to Mockoon server.
- (Ditto ability for the client to open a WebSocket connection to Mockoon server).
- Ability to define response payloads (usually JSON) that Mockoon server can send over the SSE (resp. WebSocket) connection.
- Above mentioned responses are sent with a delay of (0 .. n seconds) after the connection is opened (or even a random delay to simulate real life events).
- Above mentioned responses would allow variable contents, as with existing Mockoon responses, to simulate real life.

Signed-off-by: Andrew Fiddian-Green <software@whitebear.ch>

**Comment by isuru89** [comment]
I think this issue could get more popular with the rise of the "GPT" world. Because, those GPT APIs uses SSEs to transfer the real-time generating contents.

The reasons I see:
 1. **Popularity** of GPT APIs.
 2. **Cost of Usage:** Also, these APIs are commercialized or rate-limited. So, using actual APIs are limited in development or test environments due to the additional incurring cost to the organization. Hence, developers would expect a mocking API in their local machine/non-production environments.
 3. **Predictability**: Such GPT APIs tend to produce non-deterministic results even with the same inputs. So, sometime troubleshooting a scenario would not be easy and may require a help from mocking API to simulate with such troublesome responses and behaviors.

**Comment by yunqiangwu** [comment]
我也想要这种功能

**Comment by dagadbm** [comment]
it kind of works as is but for some reason i always end the messages with an error..

**Comment by 255kb** [comment]
> it kind of works as is but for some reason i always end the messages with an error..

Could you elaborate on how you currently simulate SSE?

**Comment by dagadbm** [comment]
just hit record on the proxy and run your app. you will get the mock done
automatically.

**Comment by Jdruwe** [comment]
Just wondering, is this feature planned to be developed?

**Comment by 255kb** [comment]
@Jdruwe Not yet. You can find our (indicative) roadmap [here](https://github.com/orgs/mockoon/projects/9/views/1). 
Maybe @isuru89 wants to tackle this subject? We already have a beginning of an implementation with the Admin API events SSE endpoint.

**Comment by itstueben** [comment]
Would be great to make the SSE Feature avalible in mockoon.

**Comment by isuru89** [comment]
@255kb 
Thank you for reminding me this.
Sometime back I was working on this, but then prioritized over Websocket implementation.

I will have a look as soon as I find a free time during next week.

**Comment by 255kb** [comment]
> [@255kb](https://github.com/255kb) Thank you for reminding me this. Sometime back I was working on this, but then prioritized over Websocket implementation.
> 
> I will have a look as soon as I find a free time during next week.

Feel free to reach out on Discord if you want to discuss. Adding this feature would require a major release which is not planned before later this year. 
You will see in the commons-server lib that there is already an [sse.ts](https://github.com/mockoon/mockoon/blob/ce6bb58a0b59a00945d342158981cd1b27131932/packages/commons-server/src/libs/server/sse.ts) implementation which was used for the admin API `/events` endpoint (not documented yet). 
I think the implementation could be similar to websockets, where you can set up some body and send it every x ms. I don't know if the distinction between unicast and broadcast would make sense, I would say not. It would keep the implementation simpler.

**Comment by isuru89** [comment]
@255kb Thank you for the info. I tried to reach you in Discord, but does not allow me to send messages to you. So, until then I will specify here one of the PoC I carried out some months back.

My idea is like this. I found my piece of code last week and I pushed into my fork. https://github.com/isuru89/mockoon/pull/2/files

-  There will not be any new entry in the Mainmenu like for creating WebSocket. It will be a variation of the existing HTTP GET endpoint.
- When a user creates a HTTP endpoint and selects the GET method, the two options will be shown like "None" mode and "Event Stream" mode. We could call this mode internally "AliveMode".
- User cannot specify SSE for any CRUD GET endpoints. Only  available in HTTP endpoints.
- Existing GET endpoints will not be affected, because default alive mode will be "None".
- We borrow the similar set of options as in Websockets. Interval will indicate the gap between two consecutive event response from the stream.
- Response modes, Rules and Settings will be supported as it is.
- Custom headers defined in the "Headers" tab will be sent initially, but we will not allow to override Content-Type and other required headers from there.
- All the existing tabs will be shown except for "Callback" tab. I personally don't see any usecase for callbacks.
- How do we define a streaming responses?
  1 If user has selected File type, we will read the file and stream the content of the file line-by-line. Templating will be enabled/disabled based on the settings.
     2. Similary if user has selected "Inline" type, the content will be served similary line-by-line. 
     3. If the "Databucket" type is selected, one of followings will be activated.
          - If the content is json:
             - If it is an array, each array entry will be streamed separately
             - If it is an object, the whole object will be streamed once and response will be closed
          - Otherwise, the content will be streamed line-by-line.

What I have not thought so far is:
  * How should a log entry should look like for SSE endpoint? Should we show all response lines? (Or trim it to the last N lines served?)
  * Proxy mode will not be supported (right?)
  * Obviously, icons and naming 

Would love to hear any feedback/thoughts around this type.

**Comment by 255kb** [comment]
@isuru89 Sorry for getting back to you so late. This topic completely slipped my mind. 

I agree that HTTP and SSE are nearly identical, but I think we need a new entry for feature discovery and clarity. At some point I was even thinking that we could trigger SSE behavior by creating some new templating helper like `{{# stream}}`. Not sure if it was a good idea but maybe it's worth exploring? It would allow us to add the feature without a major version bump. And we could still add a menu entry creating this specific template. 

For the rest I think we are on the same page, except maybe:
- the callbacks: I don't see why someone wouldn't want a callback to be sent after opening the SSE connection (probably an edge case, true). 
- the streaming: why the difference between inline and data buckets? I get that data buckets are designed to contain JSON by default. I also have mixed feeling about the streaming line by line when it's inline mode, but full objects at once when it's a data bucket. And what about someone wanting to always send the same object, like an event, but because it's inline it gets regenerated everytime? 

I think we should list very specific use cases (sending events of the same object, OpenAI API stream mode, etc.) and start from there. Maybe also implement only one or two at first. 
Having SSE created from a template helper would also let us create more helpers or helper params that could define how the content would be streamed: line by line, everything every time, objects in an array, etc.

**Comment by ahaoli** [comment]
Just add data: message\n\n, set the Content-Type to text/event-stream, and it works. But there’s no delay between the two SSE lines.

**Comment by 255kb** [comment]
> Just add data: message\n\n, set the Content-Type to text/event-stream, and it works. But there’s no delay between the two SSE lines.

Aren't the lines sent all at once and the connection immediately closed like with a regular HTTP request? I don't think Mockoon is currently capable of simulating SSE, at least from my tests.

**Comment by ahaoli** [comment]
> > Just add data: message\n\n, set the Content-Type to text/event-stream, and it works. But there’s no delay between the two SSE lines.
> 
> Aren't the lines sent all at once and the connection immediately closed like with a regular HTTP request? I don't think Mockoon is currently capable of simulating SSE, at least from my tests.

yes, it's sended by mockoon line by line. (so quick that human can not make sense, but i use java code and debug it. it's real sse stream, here is the screen shots you can see)

<img width="1524" height="236" alt="Image" src="https://github.com/user-attachments/assets/ec487151-fa70-4bbe-8323-11128fc6a2d2" />

<img width="739" height="446" alt="Image" src="https://github.com/user-attachments/assets/8e677ed6-c0fe-4ba2-a3e3-ff89ea2d6fbe" />

<img width="1249" height="429" alt="Image" src="https://github.com/user-attachments/assets/70bda58a-f7e5-4535-affb-2a13b346d11b" />

**Comment by 255kb** [comment]
@ahaoli thanks for the precision, that's interesting. 
Then, this feature could probably be implemented quickly by detecting the relevant SSE headers and delaying the lines, by the already available delay. 
After that, specific helpers could be introduced to account for more use cases like repeating the same message over time.
